File: LeetCodePractice/src/twoPointers/window/LongestSubstringWithoutRepeatingCharacters.java
package twoPointers.window;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/*
Given a string, find the length of the longest substring without repeating characters.

Example 1:

Input: "abcabcbb"
Output: 3
Explanation: The answer is "abc", with the length of 3.
Example 2:

Input: "bbbbb"
Output: 1
Explanation: The answer is "b", with the length of 1.
Example 3:

Input: "pwwkew"
Output: 3
Explanation: The answer is "wke", with the length of 3.
             Note that the answer must be a substring, "pwke" is a subsequence and not a substring.
 */
public class LongestSubstringWithoutRepeatingCharacters {
	/**
	 * @param s:
	 *            a string
	 * @return: an integer
	 */
	public int lengthOfLongestSubstring(String s) {
		int res = 0, len = s.length(), start = 0, end = 0;
		Set<Character> set = new HashSet<>();
		while(end < len){
			char ec = s.charAt(end);
			while(set.contains(ec)){
				char sc = s.charAt(start);
				set.remove(sc);
				start++;
			}
			set.add(ec);
			end++;
			res = Math.max(res, end - start);
		}
		return res;
	}

	public int lengthOfLongestSubstringII(String s) {
		if(s == null || s.length() == 0){
			return 0;
		}
		int res = 0, len = s.length(), start = 0, end = 0;
		Map<Character, Integer> map = new HashMap<>();
		while(end < len){
			char ec = s.charAt(end);
			if(map.containsKey(ec)){
				if(start < map.get(ec) + 1){
					start = map.get(ec) + 1;
				}
			}
			res = Math.max(res, end - start + 1);
			map.put(ec, end);
			end++;
		}
		return res;
	}
}

File: LeetCodePython/LongestSubstringWithoutRepeatingCharacters.py
"""
Given a string s, find the length of the longest substring without repeating characters.

Example 1:
Input: s = "abcabcbb"
Output: 3
Explanation: The answer is "abc", with the length of 3.

Example 2:
Input: s = "bbbbb"
Output: 1
Explanation: The answer is "b", with the length of 1.

Example 3:
Input: s = "pwwkew"
Output: 3
Explanation: The answer is "wke", with the length of 3.
Notice that the answer must be a substring, "pwke" is a subsequence and not a substring.


Constraints:
0 <= s.length <= 5 * 10^4
s consists of English letters, digits, symbols and spaces.
"""


class LongestSubstringWithoutRepeatingCharacters:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res, l, r = 0, 0, 0
        char_set = set()
        while r < len(s):
            while s[r] in char_set:
                char_set.remove(s[l])
                l += 1
            char_set.add(s[r])
            r += 1
            res = max(res, r - l)
        return res
